### PR TITLE
ROX-21669: Utilize local mapping files

### PR DESCRIPTION
**Pull request by daynewlee**
## Description
Deal with the situation when indexer is ready before Central

## Checklist
- [x] Investigated and inspected CI test results

## Manual test
1.
```
make cli-darwin_amd64

./bin/darwin_amd64/roxctl helm output central-services --image-defaults=development_build --debug-path=./image --debug --remove

export KUBECONFIG=<kubeconfig path from downloaded infra artifacts>

vals_dir=<path to dir holding helm values files>

```

2. set "$vals_dir/secret_values.yaml"
```
imagePullSecrets:
   username: "<your user>"
   password: "<your pass>"
```

3. set "$vals_dir/central_values.yaml"
```
enablePodSecurityPolicies: false
allowNonstandardNamespace: true
customize:
  envVars:
    LOGLEVEL: "Debug"
image:
  registry: "quay.io"
central:
  image:
    tag: "4.3.x-679-g5e0240dea1"
    name: "rhacs-eng/main"
  db:
    enabled: true
    image:
      tag: "4.3.x-679-g5e0240dea1"
      name: "rhacs-eng/central-db"
    resources:
      requests:
        cpu: 2
        memory: "4Gi"
      limits:
        cpu: 2
        memory: "4Gi"
  persistence:
    none: true
  exposure:
    loadBalancer:
      enabled: true
  telemetry:
    enabled: false
scanner:
  disable: false
  replicas: 1
  autoscaling:
    minReplicas: 1
    maxReplicas: 1
  image:
    tag: "4.3.x-679-g5e0240dea1"
    name: "rhacs-eng/scanner"
  dbImage:
    tag: "4.3.x-679-g5e0240dea1"
    name: "rhacs-eng/scanner-db"
scannerV4:
  disable: false
  indexer:
    image:
      tag: "4.3.x-679-g5e0240dea1"
      name: "rhacs-eng/scanner-v4"
    replicas: 1
    autoscaling:
      minReplicas: 1
      maxReplicas: 1
  matcher:
    image:
      tag: "4.3.x-679-g5e0240dea1"
      name: "rhacs-eng/scanner-v4"
    replicas: 1
    autoscaling:
      minReplicas: 1
      maxReplicas: 1
    resources:
      requests:
        cpu: "1"
        memory: "1500Mi"
      limits:
        cpu: "2"
        memory: "8Gi"
  db:
    image:
      tag: "4.3.x-679-g5e0240dea1"
      name: "rhacs-eng/scanner-v4-db"
    persistence:
      persistentVolumeClaim:
        storageClass: "ssd-csi"
        size: "50Gi"
    resources:
      requests:
        cpu: "200m"
        memory: "200Mi"
      limits:
        cpu: "4"
        memory: "6Gi"
```

4. run
```
helm upgrade --install -n stackrox stackrox-central-services --create-namespace ./stackrox-central-services-chart \
        -f "$vals_dir/central_values.yaml" -f "$vals_dir/secret_values.yaml"
```

Results: **_No error found_**
```
{"level":"info","host":"scanner-v4-indexer-5d5d8bdd95-h87hl","component":"main","version":"4.3.x-731-gc9da4582cc","time":"2024-01-15T21:56:07Z","message":"starting scanner"}
pkg/metrics: 2024/01/15 21:56:07.440567 server.go:142: Warn: Secure metrics server is disabled
pkg/metrics: 2024/01/15 21:56:07.440688 server.go:142: Warn: Secure metrics server is disabled
pkg/env: 2024/01/15 21:56:07.440809 metrics.go:47: Debug: ROX_METRICS_PORT=:9090, resolved to :9090
pkg/metrics: 2024/01/15 21:56:07.440863 server.go:142: Warn: Secure metrics server is disabled
{"level":"info","host":"scanner-v4-indexer-5d5d8bdd95-h87hl","component":"main","time":"2024-01-15T21:56:07Z","message":"indexer is enabled"}
pkg/metrics: 2024/01/15 21:56:07.442543 throttle.go:90: Info: gathering CPU throttle metrics from sys/fs/cgroup/cpu.stat
{"level":"info","host":"scanner-v4-indexer-5d5d8bdd95-h87hl","component":"libindex/New","time":"2024-01-15T21:56:08Z","message":"registered configured scanners"}
{"level":"info","host":"scanner-v4-indexer-5d5d8bdd95-h87hl","component":"indexer.NewLayerScanner","time":"2024-01-15T21:56:08Z","message":"NewLayerScanner: constructing a new layer-scanner"}
{"level":"info","host":"scanner-v4-indexer-5d5d8bdd95-h87hl","component":"main","time":"2024-01-15T21:56:08Z","message":"matcher is disabled"}
{"level":"info","host":"scanner-v4-indexer-5d5d8bdd95-h87hl","component":"main","time":"2024-01-15T21:56:08Z","message":"backends are ready"}
grpc_internal: 2024/01/15 21:56:08.466744 logging.go:35: Debug: [core][Server #1] Server created
pkg/grpc: 2024/01/15 21:56:08.466975 server.go:283: Info: Launching backend gRPC listener
grpc_internal: 2024/01/15 21:56:08.467053 clientconn.go:304: Debug: [core][Channel #2] Channel created
grpc_internal: 2024/01/15 21:56:08.467070 logging.go:43: Debug: [core][Channel #2] original dial target is: ""
grpc_internal: 2024/01/15 21:56:08.467109 logging.go:43: Debug: [core][Channel #2] parsed dial target is: resolver.Target{URL:url.URL{Scheme:"", Opaque:"", User:(*url.Userinfo)(nil), Host:"", Path:"", RawPath:"", OmitHost:false, ForceQuery:false, RawQuery:"", Fragment:"", RawFragment:""}}
grpc_internal: 2024/01/15 21:56:08.467134 logging.go:43: Debug: [core][Channel #2] fallback to scheme "passthrough"
grpc_internal: 2024/01/15 21:56:08.467166 logging.go:43: Debug: [core][Channel #2] parsed dial target is: passthrough:///
grpc_internal: 2024/01/15 21:56:08.467245 logging.go:43: Debug: [core][Channel #2] Channel authority set to ""
grpc_internal: 2024/01/15 21:56:08.467234 logging.go:35: Debug: [core][Server #1 ListenSocket #3] ListenSocket created
```

### CODE REVIEW

File: scanner/indexer/indexer.go [hunk 1]
```diff
@@ -97,12 +97,22 @@ func newLibindex(ctx context.Context, indexerCfg config.IndexerConfig, store cci
...
+	if err == nil {
+		opts.ScannerConfig.Repo = map[string]func(interface{}) error{
+			"rhel-repository-scanner": deserializeFunc([]byte(fmt.Sprintf(`{"repo2cpe_mapping_url": "%s"}`, indexerCfg.RepositoryToCPEURL)))}
+		opts.ScannerConfig.Package = map[string]func(interface{}) error{
+			"rhel_containerscanner": deserializeFunc([]byte(fmt.Sprintf(`{"name2repos_mapping_url": "%s"}`, indexerCfg.NameToCPEURL)))}
+	} else { // if central or sensor is not available
+		zlog.Info(ctx).Err(err).Msg("Mapping file endpoints are currently unavailable")
+		opts.ScannerConfig.Repo = map[string]func(interface{}) error{
+			"rhel-repository-scanner": deserializeFunc([]byte(fmt.Sprintf(
+				`{"repo2cpe_mapping_url": "%s","repo2cpe_mapping_file": "%s"}`, indexerCfg.RepositoryToCPEURL, "mappingfiles/repository-to-cpe.json"))),
```
Comment (jvdm) [inline scanner/indexer/indexer.go:110]: Just to confirm: After the initial pull, the next one will reach out to the URL, correct?

File: scanner/indexer/indexer.go [hunk 2]
```diff
@@ -97,12 +97,22 @@ func newLibindex(ctx context.Context, indexerCfg config.IndexerConfig, store cci
 		ScanLockRetry:        libindex.DefaultScanLockRetry,
 		LayerScanConcurrency: libindex.DefaultLayerScanConcurrency,
 	}
-
-	opts.ScannerConfig.Repo = map[string]func(interface{}) error{
-		"rhel-repository-scanner": deserializeFunc([]byte(fmt.Sprintf(`{"repo2cpe_mapping_url": "%s"}`, indexerCfg.RepositoryToCPEURL))),
-	}
-	opts.ScannerConfig.Package = map[string]func(interface{}) error{
-		"rhel_containerscanner": deserializeFunc([]byte(fmt.Sprintf(`{"name2repos_mapping_url": "%s"}`, indexerCfg.NameToCPEURL))),
+	err = sanityCheckURL(indexerCfg.RepositoryToCPEURL)
```
Comment (jvdm) [inline scanner/indexer/indexer.go:100]: Let's consider doing this regardless of whether Central/Sensor is available; you might get the file in the sanity check, but something might fail in the actual pull, and things will be stale for 24 hours if I understood the updater logic correctly.

File: tools/allowed-large-files [hunk 3]
```diff
@@ -40,6 +40,8 @@ roxctl/helm/derivelocalvalues/testdata/k8s-resources/*.yaml
 scale/image/fixtures/scrape.json
 scanner/e2etests/testdata/image_tests.json
 scanner/e2etests/testdata/image_tests.orig.json
+scanner/indexer/mappingfiles/container-name-repos-map.json
+scanner/indexer/mappingfiles/repository-to-cpe.json
```
Comment (jvdm) [inline tools/allowed-large-files:44]: Can we create an issue to add these files to the container image instead and pull them during the image build?

File: scanner/indexer/indexer.go (hunk 1 above)
Comment (RTann) [reply scanner/indexer/indexer.go:110]: yep: https://github.com/quay/claircore/blob/main/rhel/repositoryscanner.go#L151

**Review by RTann — COMMENTED** [review]
I agree with Victor. Let's pull the files upon image build and let's always set the file in the RHEL configuration

File: scanner/indexer/indexer.go (hunk 2 above)
Comment (daynewlee) [reply scanner/indexer/indexer.go:100]: Are suggesting we can move this check to somewhere else? Do you have any recommendations?

File: tools/allowed-large-files (hunk 3 above)
Comment (daynewlee) [reply tools/allowed-large-files:44]: > add these files to the container image instead and pull them during the image build

I am not sure how to do that. Is there any examples to look at?

Comment (jvdm) [reply tools/allowed-large-files:44]: Just create the JIRA ticket so we can do it as soon as we can.

File: scanner/indexer/indexer.go (hunk 2 above)
Comment (jvdm) [reply scanner/indexer/indexer.go:100]: No, I'm suggesting we should not even do the check, but simply use the initial files, always.

File: scanner/indexer/indexer.go (hunk 1 above)
Comment (jvdm) [reply scanner/indexer/indexer.go:110]: Thanks.

File: tools/allowed-large-files (hunk 3 above)
Comment (daynewlee) [reply tools/allowed-large-files:44]: https://issues.redhat.com/browse/ROX-21690

cc: @jvdm  @jvdm

Comment (RTann) [reply tools/allowed-large-files:44]: https://github.com/stackrox/stackrox/pull/9362

**Review by RTann — COMMENTED** [review]
@jvdm @daynewlee take a look at https://github.com/stackrox/stackrox/pull/9362 which will embed the files into the image to be used by this PR

File: tools/allowed-large-files (hunk 3 above)
Comment (jvdm) [reply tools/allowed-large-files:44]: Please coordinate this change against #9362. One thing to think about is that by having the mappings in the container image, we cannot run the scanner locally anymore unless we can either ignore the files if they are not found or make their path configurable in the config.

**Review by RTann — COMMENTED** [review]
Have you been able to run tests to confirm this worked?

File: scanner/indexer/indexer.go
Comment (jvdm) [inline scanner/indexer/indexer.go:1]: Can you provide some testing information about this? In particular, does this work when running locally without the Docker Image environment?

Comment (daynewlee) [reply scanner/indexer/indexer.go:1]: I wasn't able to test this since image are not able to be uploaded

Comment (daynewlee) [reply scanner/indexer/indexer.go:1]: I will bypass the arm64 build and keep testing

File: .github/workflows/scanner-build.yml [hunk 4]
```diff
@@ -19,7 +19,7 @@ jobs:
     strategy:
       matrix:
         # If this is updated, be sure to update architectures in push-manifests below.
-        goarch: ['amd64', 'arm64', 'ppc64le', 's390x']
```
Comment (daynewlee) [inline .github/workflows/scanner-build.yml:22]: Just for testing purpose, maybe we can remove it if the team agree on this

Comment (daynewlee) [reply .github/workflows/scanner-build.yml:22]: Will rebase with master when https://github.com/stackrox/stackrox/pull/9389/files is merged

Comment (RTann) [reply .github/workflows/scanner-build.yml:22]: rebase

**Review by RTann — APPROVED** [review]
just rebase then should be good, thanks

**Pull request merged**